collections/ansible_collections/cloudkit/service/roles/hostpool_infra/tasks/create_hostpool_infra.yaml

```diff
@@ -0,0 +1,107 @@
+- name: Show host requests
+  ansible.builtin.debug:
+    var: hostpool_infra_host_requests
+
+- name: Create hostpool network
+  ansible.builtin.include_role:
+    name: massopencloud.esi.network
+  vars:
+    network_suffix: "{{ hostpool_infra_name }}"
+    network_state: present
+    network_properties:
+      instance: "{{ hostpool_infra_namespace }}"
+      hostpool: "{{ hostpool_infra_name }}"
+      purpose: "o-sac"
+
+- name: Set hostpool_infra_network_name
+  ansible.builtin.set_fact:
+    hostpool_infra_network_name: "network-{{ hostpool_infra_name }}"
+
+# Handle scale down **before** scale up
+
+- name: List host classes currently in use by hostpool
+  ansible.builtin.include_role:
+    name: massopencloud.esi.host
+    tasks_from: list_host_classes_for_hostpool
+  vars:
+    host_pool_name: "{{ hostpool_infra_name }}"
+
+- name: Set host classes included in the request
+  ansible.builtin.set_fact:
+    requested_host_classes: >
+      {{ hostpool_infra_host_requests | map(attribute='hostClass') }}
+
+- name: Determine which host classes are no longer needed
+  ansible.builtin.set_fact:
+    removed_host_classes: "{{ removed_host_classes | default([]) + [item] }}"
+  with_items: "{{ host_class_list }}"
+  when: item not in requested_host_classes
+
+- name: Display host classes to be removed
+  ansible.builtin.debug:
+    msg: "The following host classes are no longer requested: {{ removed_host_classes | default([]) }}"
+
+- name: Initialize updated hostpool host requests
+  ansible.builtin.set_fact:
+    updated_hostpool_infra_host_requests: "{{ hostpool_infra_host_requests }}"
+
+- name: Add removed host classes to updated hostpool host requests
+  ansible.builtin.set_fact:
+    updated_hostpool_infra_host_requests: "{{ updated_hostpool_infra_host_requests + [{'size': '0', 'hostClass': item}] }}"
+  with_items: "{{ removed_host_classes | default([]) }}"
+
+- name: Remove unneeded hosts from hostpool
+  ansible.builtin.include_role:
+    name: manage_hosts
+    tasks_from: remove_unneeded_hosts
+  vars:
+    manage_hosts_hostpool_name: "{{ hostpool_infra_name }}"
+    manage_hosts_desired_count: "{{ item.size | int }}"
+    manage_hosts_host_class: "{{ item.hostClass }}"
+  loop: "{{ updated_hostpool_infra_host_requests }}"
+
+- name: Add new hosts to hostpool
+  ansible.builtin.include_role:
+    name: manage_hosts
+    tasks_from: label_new_hosts
+  vars:
+    manage_hosts_hostpool_name: "{{ hostpool_infra_name }}"
+    manage_hosts_desired_count: "{{ item.size | int }}"
+    manage_hosts_host_class: "{{ item.hostClass }}"
+  loop: "{{ hostpool_infra_host_requests }}"
+
+- name: List hosts currently labeled with the hostpool label
+  ansible.builtin.include_role:
+    name: massopencloud.esi.host
```

Comment:
Here's something that we can fix later:

We're using the `massopen.esi` collection to translate OSAC concepts into ESI concepts. This seems like the wrong layer at which to do this, because it might make it more difficult to swap out ESI for some other inventory mechanism.

It seems like we want some a layer in between this playbook and `massopen.esi`; that layer should export a stable inventory to callers (with operations like `list_hosts`, `add_host_to_group`, `remove_host_from_group`, etc) and then translate those operations into implementation-specific actions.

In that world, maybe we would do something like:

```
ansible.builtin.include_role:
  name: "{{ osac_inventory_driver }}.host"
  tasks_from: list_hosts
```

Etc.